Add drag and drop functionatliy to input elements

dcdenu4 opened the pull request:
This PR allows for the dragging and dropping of files into input text boxes. It adds the following:

- When dragging over a non `input` tag on the SetupTab component set `opacity` to 0.2 to indicate `datastack` drop
- When dragging over an `input` tag on the SetupTab component add dashed border to input for emphasis.
- On drag leave of `input` tag remove dashed border emphasis.

Tests:
------
I added tests but for the life of me could not figure out why `drop` on the `input` test kept failing. Here's what I think is going wrong:
- `ArgsForm.inputDropHandler()` expects `event.dataTransfer.files` to return a `FileList` Object from `input` elements.
- https://developer.mozilla.org/en-US/docs/Web/API/FileList
- However, only a string Object is returned when testing and thus does not have a `path` attribute.
- `path` attribute returns `Undefined`
- To test uncomment the following in `ArgsForm.inputDropHandler()` and view from test output and from workbench output:
    - `console.log(fileList);`
    - `console.log(fileList[0]);`
    - `console.log(fileList[0].path);`
 
Test output:
```
[ 'file-path']
[file-path]
undefined
```
Workbench output:
```
FileList {0: File, length: 1}
File {name: "whatever file you dropped in", path: "path to file dropped"}
Path to file dropped
```

Fixes #37

emlys reviewed (changes requested):
Hey @dcdenu4, this looks really cool! I tried out the dragging-and-dropping and it's very fun. A few things I noticed:
-  Dropping a file onto a disabled input works, I would expect it not to
- When I drag a file across the empty space in the sidebar (below 'FAQ'), the file icon on my cursor gets a little green plus icon. It looks like it's saying something should happen when I drop it there. It doesn't have that icon over the form area, or in other apps that accept drag-and-drop. Very weirdly, the icon disappears when I press cmd+option+3 to take a screenshot of it, so I can't share a picture. If this sounds like a mac-only thing I'd be happy to help with it.
- The alert message when you drop an invalid file onto the form could maybe be clearer? It does make sense but I could see users being confused by it. Especially when the file is not a datastack file, it says "Parameter/Log file for undefined..." Maybe something like, "This file is not a valid datastack for this model."

dcdenu4 commented:
Hey @davemfish and @emlys I went through and did a lot more work on the drag / drop feature. 

Emily, I *think* I addressed all the comments you made. Hopefully the interaction is more clear from a user perspective. 

Please clone the branch and test it out! 

I did add lots of `dragOverHandler` functions to various modules. I was torn on whether each component should handle it's own actions or if passing down from some on high parent would make more sense. Ultimately I thought that each component should handle it's own.

emlys commented:
@dcdenu4 I'm still getting the [copy cursor](https://support.apple.com/en-om/guide/mac-help/mh35695/11.0/mac/11.0) on the Home tab and bordering areas that aren't within a react component. This may be an electron issue, it happens in the [quick start app](https://www.electronjs.org/docs/tutorial/quick-start) too. Googling didn't turn up anything similar, so maybe it's just me? (@phargogh would you mind trying this out sometime?)

dcdenu4 commented:
> I'm still getting the [copy cursor](https://support.apple.com/en-om/guide/mac-help/mh35695/11.0/mac/11.0) on the Home tab and bordering areas that aren't within a react component

Hey @emlys I just pushed up some changes that should display the `none` symbol when dragging over anything but the very outside border, which is simply the `<body>` tag ( on Windows anyway ). It seems to me that the default behavior for dragging something over an application is the `copy` symbol, even though the element has no `dragover` event handler. By default dragging over and dropping something does nothing unless those defaults are prevented.

I do think that it looks better to have the `none` symbol pop up over nonsensical elements.

davemfish commented on tests/setuptab.test.js:
```diff
@@ -580,4 +587,260 @@ describe('Form drag-and-drop', () => {
+    expect(setupInput).not.toHaveClass("input-dragging");
+  });
+
+  test('Drag and drop on a disabled input element.', async () => {
```
I thought the intent here would be to do the drop and expect no updated value in the input, but this test does a "drag & leave" with no drop.

dcdenu4 replied:
Good catch! Fixed here: https://github.com/natcap/invest-workbench/pull/92/commits/744114de22a529559e2d1273d1697074bbf05b30

davemfish reviewed (changes requested):
Hey @dcdenu4 , I tested this and it works great! I really like the visual cues.

I've got some misc comments - mostly linting type stuff and one test that I think is only passing by coincidence. 

It sure would be nice to have a way to change that default `dropEffect` from 'copy' to 'none' for all the components that don't support drag/drop. I don't love having to add the handlers to all those unrelated components, but alas I have no other solution to offer 🤷

phargogh commented:
Hey, sorry for the delay on following up here @emlys 

> @dcdenu4 I'm still getting the copy cursor on the Home tab and bordering areas that aren't within a react component. This may be an electron issue, it happens in the quick start app too. Googling didn't turn up anything similar, so maybe it's just me? (@phargogh would you mind trying this out sometime?)

When I try it out, I'm getting the copy cursor initially for the whole model inputs pane (expected for drag-n-drop for datastacks/logfiles), but then the cursor reverts to a normal arrow after holding position for a few seconds.

I'm also seeing the drag-n-drop highlighting for a single input, and that works great!  The visual queue there is really nice and clear, too.

emlys commented on tests/setuptab.test.js:
```diff
@@ -580,4 +587,260 @@ describe('Form drag-and-drop', () => {
+        arg1: {
+          name: 'Workspace',
+          type: 'directory',
+          disabled: true,
```
Yep, after #84 you can pass a boolean `enabled` prop to `ArgInput`s. For a couple of tests I mocked out the UI spec file like this
```
beforeAll(() => {
    const mockUISpec = {FooModel: {order: ["foo", "bar"]}};
    jest.mock('../src/ui_config', () => mockUISpec);
}
afterAll(() => {
    jest.resetModules();
}
```

dcdenu4 replied:
What was I looking at? 🙃

dcdenu4 replied:
I ultimately went with using the current `uiSpec` setup, knowing that it will change... I did not go with rendering `ArgInput` because it relies on handlers from `ArgsForm`. https://github.com/natcap/invest-workbench/pull/92/commits/744114de22a529559e2d1273d1697074bbf05b30

emlys commented:
Agree with @davemfish's comments! I opened an issue for the copy cursor thing: https://github.com/electron/electron/issues/27594 and I think the way it is is fine for now.

dcdenu4 commented on src/components/SetupTab/ArgsForm/index.jsx:
```diff
@@ -8,40 +8,98 @@ import ArgInput from '../ArgInput';
     } else {
-      throw alert(`Parameter/Log file for ${datastack.module_name} does not match this model: ${this.props.pyModuleName}`)
+      alert(
+        "Cannot update inputs with dropped file. Please use saved datastack files or model logfiles.")
```
That's a good point. I see that the original message makes much more sense now. I'll revert it back.

dcdenu4 replied:
Updated back here: https://github.com/natcap/invest-workbench/pull/92/commits/88b2951c682fa241ac4fd9f6cda6445c8780bf22

dcdenu4 commented on src/app.jsx:
```diff
@@ -19,6 +19,13 @@ import InvestJob from './InvestJob';
+  event.preventDefault();
+  event.stopPropagation();
+  event.dataTransfer.dropEffect = 'none';
+}
```
This is what I've encountered in practice. I tried looking more into it but couldn't actually confirm default behavior for `dropEffect` where elements don't have a `dragOver` event handler. I was telling Emily that we could write a `DragDrop` component that we then use, but it's kind of the same problem of including it everywhere. For the `dragOverHandler` particularly, we could put it in utils and import the function when needed...

davemfish replied:
Yeah, throwing it in `utils` is an option if you want to. That's pretty much what `utils` is there for!

dcdenu4 replied:
Moved to utils here: https://github.com/natcap/invest-workbench/pull/92/commits/cc7c7a8f44be532dba11313d56e4707d21401f05

dcdenu4 commented on src/components/SetupTab/ArgsForm/index.jsx:
```diff
@@ -8,40 +8,98 @@ import ArgInput from '../ArgInput';
+    event.target.classList.remove("input-dragging");
+    // Don't take any action on disabled elements
+    if (event.target.disabled) {
+       return;
```
Yes there is. On my list todo is get my Vim editor set up for proper `jsx` editing.

dcdenu4 commented on src/components/SetupTab/ArgsForm/index.jsx:
```diff
@@ -8,40 +8,98 @@ import ArgInput from '../ArgInput';
 
   handleChange(event) {
-    /** Pass input value up to SetupTab for storage & validation. */
+    /** Pass input value up to setuptab for storage & validation. */
```
Funny story, I mashed some keys in Vim on accident and later found out that everything was set to lowercase that was capitalized. Caused quite the headache when things stopped working.

emlys commented:
Looks like the electron folks agree that the copy cursor is unexpected behavior, so hopefully we can remove those extra `onDragOver` handlers in a future release.

dcdenu4 commented:
Okay @emlys @davemfish I think I covered everything mentioned. And it looks like we'll see how Electron deals with that default behavior. Let me know what ya'll think!

emlys commented on src/components/SetupTab/SetupButtons/index.jsx:
```diff
@@ -4,6 +4,13 @@ import Button from 'react-bootstrap/Button';
 import SaveFileButton from '../../SaveFileButton';
 
+/** Prevent the default case for onDragOver so onDrop event will be fired. */
+function dragOverHandler(event) {
```
Could this also reference the `dragOverHandlerNone` in utils?

davemfish reviewed (changes requested):
Putting that function in `utils` definitely feels like a good call! I think I found one more use-case for that. And one tiny bug in a test. Back to you @dcdenu4

emlys commented on src/components/SetupTab/ArgsForm/index.jsx:
```diff
@@ -8,35 +8,93 @@ import ArgInput from '../ArgInput';
     } else {
-      throw alert(`Parameter/Log file for ${datastack.module_name} does not match this model: ${this.props.pyModuleName}`)
+      alert(
+        `Parameter/Log file for ${datastack.module_name} does not match this model: ${this.props.pyModuleName}`);
```
I'm still not crazy about this message, but it's fine to wait until we do more UX-related stuff.

dcdenu4 replied:
I know! Your prior comment made me change it, then Dave's comment made me change it back, because it is very descriptive of "how did I end up here?". But yea, we can sort some of these UX related things out in time!

emlys reviewed (commented):
Looks good @dcdenu4! Just a minor comment.

emlys commented on src/components/SetupTab/SetupButtons/index.jsx:
```diff
@@ -29,6 +36,7 @@ export function RunButton(props) {
       size="lg"
       onClick={props.wrapInvestExecute}
       disabled={props.disabled}
+      onDragOver={dragOverHandler}
```
you beat me to it!

dcdenu4 replied:
Okay okay, I forgot on. Swap made here: https://github.com/natcap/invest-workbench/pull/92/commits/5234c384637ee69313695a0ace8b911ebc1dfa83

davemfish commented on tests/setuptab.test.js:
```diff
@@ -580,4 +588,266 @@ describe('Form drag-and-drop', () => {
+    fireEvent(setupInput, fileDropEvent);
+
+    expect(setupInput).not.toHaveClass("input-dragging");
+    expect(setupInput).not.toHaveValue("foo.txt");
```
Or maybe even better than expecting "not the value" would be expecting no value? (can't remember if it would be undefined or `''`).

dcdenu4 replied:
That's smart! Now testing for empty string here: https://github.com/natcap/invest-workbench/pull/92/commits/8603806a780bf7af076df30ab68b9518f9441fb8

dcdenu4 commented:
Alright! I won't even say I think I handled everything haha. Back to you all!

emlys reviewed (approved):
Looks good to me @dcdenu4!

davemfish reviewed (approved):
Hooray! I'll have a fix for the flaky tests soon, I promise.
